src/test/net_tests.cpp

```diff
@@ -926,4 +1002,59 @@ BOOST_AUTO_TEST_CASE(node_eviction_test)
     }
 }
 
+BOOST_AUTO_TEST_CASE(initial_advertise_from_version_message)
+{
+    // Pretend that we bound to this port.
+    const uint16_t bind_port = 20001;
+    m_node.args->ForceSetArg("-bind", strprintf("3.4.5.6:%u", bind_port));
+
+    // Our address:port as seen from the peer - 2.3.4.5:20002 (different from the above).
+    in_addr peer_us_addr;
+    peer_us_addr.s_addr = htonl(0x02030405);
+    const CAddress peer_us{CService{peer_us_addr, 20002}, NODE_NETWORK};
+
+    // Create a peer with a routable IPv4 address.
+    in_addr peer_in_addr;
+    peer_in_addr.s_addr = htonl(0x01020304);
+    CNode peer{0,
+               NODE_NETWORK,
+               0,
+               INVALID_SOCKET,
+               CAddress{CService{peer_in_addr, 8333}, NODE_NETWORK},
+               0,
+               0,
+               CAddress{},
+               std::string{},
+               ConnectionType::OUTBOUND_FULL_RELAY};
+
+    const uint64_t services{NODE_NETWORK | NODE_WITNESS};
+    const int64_t time{0};
+    // Our address, as seen from the peer.
+    const CNetMsgMaker msg_maker{PROTOCOL_VERSION};
+    const auto msg = msg_maker.Make(NetMsgType::VERSION, PROTOCOL_VERSION, services, time, peer_us);
+
+    CDataStream recv_stream{msg.data, SER_NETWORK, PROTOCOL_VERSION};
+
+    // Force CChainState::IsInitialBlockDownload() to return false.
+    // Otherwise PushAddress() isn't called by PeerManager::ProcessMessage().
+    SetMockTime(1);
+    BOOST_REQUIRE(!m_node.chainman->ActiveChainstate().IsInitialBlockDownload());
+    SetMockTime(0);
```

Comment:
nit: Might be possible to write

```suggestion
TestChainState& chainstate = *(TestChainState*)&m_node.chainman->ActiveChainstate();
chainstate.JumpOutOfIbd();
```

Reply:
Done. This cast to a class which it actually is not looks quite hackish.